subsys/bluetooth/services/bms_srv.c

```diff
@@ -0,0 +1,434 @@
+
+		bt_conn_unref(conn);
+	} else {
+		LOG_INF("Removing bond information for: %s", addr_str);
+
+		err = bt_unpair(BT_ID_DEFAULT, addr);
+		if (err) {
+			LOG_ERR("Unable to remove bond information: %d", err);
+		}
+
+		memset(addr, 0, sizeof(*addr));
+	}
+}
+
+static void bond_items_remove(void)
+{
+	for (uint32_t i = 0; i < ARRAY_SIZE(bms_inst.bonded_addrs); i++) {
+		bond_item_remove(&bms_inst.bonded_addrs[i]);
+	}
+}
+
+static void all_bond_items_add(const struct bt_bond_info *info,
+			       void *user_data)
+{
+	bond_item_add(&info->addr);
+}
+
+static void rest_bond_items_add(const struct bt_bond_info *info,
+				void *user_data)
+{
+	const bt_addr_le_t *req_addr = (const bt_addr_le_t *) user_data;
+
+	if (bt_addr_le_cmp(&info->addr, req_addr) != 0) {
+		bond_item_add(&info->addr);
+	}
+}
+
+static void bond_add_execute(struct k_work *item)
+{
+	struct bt_gatt_bms_srv_bond_add_exec *exec =
+		CONTAINER_OF(item, struct bt_gatt_bms_srv_bond_add_exec, work);
+
+	/* Add bonds to the storage on client's delete request. */
+	switch (exec->op_code) {
+	case BT_GATT_BMS_SRV_OP_DEL_BOND_REQ_DEVICE:
+		bond_item_add(&exec->req_addr);
+		break;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BONDS:
+		bt_foreach_bond(BT_ID_DEFAULT, all_bond_items_add, NULL);
+		break;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BUT_ACTIVE_BOND:
+		bt_foreach_bond(BT_ID_DEFAULT, rest_bond_items_add,
+				&exec->req_addr);
+		break;
+	}
+
+	/* Transition to the next state and delete bonds from disconnected
+	 * peers using a different work item.
+	 */
+	bms_inst.state = DELETION_IN_PROGRESS;
+	k_work_submit(&bms_inst.bond_del_exec);
+
+	LOG_DBG("Bond adding work is done");
+}
+
+static void bond_del_execute(struct k_work *item)
+{
+	bond_items_remove();
+	bms_inst.state = IDLE;
+
+	LOG_DBG("Bond deleting work is done");
+}
+
+static bool ctrl_pt_op_code_validate(enum bt_gatt_bms_srv_op op_code)
+{
+	switch (op_code) {
+	case BT_GATT_BMS_SRV_OP_DEL_BOND_REQ_DEVICE:
+		return bms_inst.features.delete_requesting.supported;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BONDS:
+		return bms_inst.features.delete_all.supported;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BUT_ACTIVE_BOND:
+		return bms_inst.features.delete_rest.supported;
+	default:
+		return false;
+	}
+}
+
+static bool ctrl_pt_auth_op_check(enum bt_gatt_bms_srv_op op_code)
+{
+	switch (op_code) {
+	case BT_GATT_BMS_SRV_OP_DEL_BOND_REQ_DEVICE:
+		return bms_inst.features.delete_requesting.auth;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BONDS:
+		return bms_inst.features.delete_all.auth;
+	case BT_GATT_BMS_SRV_OP_DEL_ALL_BUT_ACTIVE_BOND:
+		return bms_inst.features.delete_rest.auth;
+	default:
+		return false;
+	}
+}
+
+static bool ctrl_pt_authorize(enum bt_gatt_bms_srv_op op_code,
+			     const uint8_t *auth_code,
+			     uint16_t auth_code_len)
+{
+	if (!ctrl_pt_auth_op_check(op_code)) {
+		return true;
+	}
+
+	if (bms_inst.cbs->operation_authorize) {
+		return bms_inst.cbs->operation_authorize(auth_code,
+							auth_code_len);
+	} else {
+		return false;
```

Comment:
I think this behavior should be documented in the API.

Reply:
I will provide more detailed API docs for initialization function. However, this line should never execute because you the user cannot initialize BMS with authorized operations without providing authorize() callback (returns EINVAL).